engine/runtime/src/main/java/org/enso/interpreter/node/controlflow/caseexpr/DateTimeBranchNode.java

```diff
@@ -0,0 +1,63 @@
+package org.enso.interpreter.node.controlflow.caseexpr;
+
+import com.oracle.truffle.api.RootCallTarget;
+import com.oracle.truffle.api.dsl.Fallback;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.nodes.NodeInfo;
+import com.oracle.truffle.api.profiles.ConditionProfile;
+import org.enso.interpreter.runtime.data.EnsoDateTime;
+import org.enso.interpreter.runtime.data.Type;
+
+/** An implementation of the case expression specialised to working on Date_Time. */
+@NodeInfo(shortName = "DateMatch")
+public abstract class DateTimeBranchNode extends BranchNode {
+  private final Type dateTime;
+  private final ConditionProfile profile = ConditionProfile.createCountingProfile();
+
+  DateTimeBranchNode(Type vector, RootCallTarget branch) {
+    super(branch);
+    this.dateTime = vector;
+  }
+
+  /**
+   * Creates a new node for handling matching on a case expression.
+   *
+   * @param dateTime the expression to use for matching
+   * @param branch the expression to be executed if (@code matcher} matches
+   * @return a node for matching in a case expression
+   */
+  public static DateTimeBranchNode build(Type dateTime, RootCallTarget branch) {
+    return DateTimeBranchNodeGen.create(dateTime, branch);
+  }
+
+  @Specialization
+  void doType(VirtualFrame frame, Object state, Type target) {
+    if (profile.profile(dateTime == target)) {
+      accept(frame, state, new Object[0]);
+    }
+  }
+
+  @Specialization
+  void doEnsoDateTime(VirtualFrame frame, Object state, EnsoDateTime dateTime) {
+    accept(frame, state, new Object[0]);
+  }
+
+  @Specialization(
+      guards = {
+        "interop.isDate(dateTime)",
+        "interop.isTime(dateTime)",
```

Comment:
shouldn't we also check for timezone? Seems like (at least theoretically) `LocalDateTime` would satisfy these two checks, but I don't think it should be matched with `Date_Time` as it's not implicitly convertible to `ZonedDateTime` (what would be the timezone?)

Reply:
I don't think so. For LocalDateTime we implicitly add System's default timezone. See Time_Utils and EnsoDateTime

Reply:
Right! I forgot about it.

I'm very unconvinced that this is the right thing to do - but that is another discussion - your changes are consistent with the rest, so that's ok.